proposals/csharp-12.0/collection-expressions.md

```diff
@@ -515,31 +515,31 @@ A collection expression has a *known length* if the compile-time type of each *s
 ### Interface translation
 [interface-translation]: #interface-translation
 
-Given a target type `IEnumerable<T>`, `IReadOnlyCollection<T>`, `IReadOnlyList<T>`, `ICollection<T>`, or `IList<T>`a compliant implementation is only required to produce a value that implements that interface.  A compliant implementation is free to: 
+Given a target type `IEnumerable<T>`, `IReadOnlyCollection<T>`, `IReadOnlyList<T>`, `ICollection<T>`, or `IList<T>`, a compliant implementation is required to produce a value that implements that interface. It is recommended that any type that is synthesized implement all these interfaces. This ensures maximal compatibility with existing libraries, including those that introspect the interfaces implemented by a value in order to light up performance optimizations.
 
-1. Use an existing type that implements that interface.
-1. Synthesize a type that implements the interface.
+In addition, the value must implement the nongeneric `ICollection` and `IList` interfaces. This enables collection expressions to support dynamic introspection in scenarios such as data binding.
```

Comment:
Should this be "must implement the nongeneric ... interfaces *corresponding to the generic target type*"?

From the LDM notes below, it wasn't clear to me whether "appropriate corresponding non-generic interfaces" meant corresponding to the specific generic target interface or all of `IEnumerable`, `ICollection`, and `IList`.

> it is in fact more important that the concrete types we provide for the generic interfaces also implement the appropriate corresponding non-generic interfaces.

Reply:
![Collection interface hierarchy](https://github.com/dotnet/csharplang/assets/5833655/699353d0-6024-48c0-a077-07a8c66a5d37)

Given the collection interface hierarchy illustrated here I think it's fine if the value we create when targeting to interface types, is simply spec'd to implement IList, as the others will come along for the ride.

Reply:
I think I might have talked around your original point.. I guess what I'm really trying to say is, I think it's fine for the value we produce to implement IList in all cases where we are converting collection-expr to generic collection interface type.

Reply:
Functionality-wise, there is very little correspondence between the generic and nongeneric hierarchies in how they define 'collection' versus 'list.'

I do think targeting IEnumerable<T> should implement IList for the data binding reason.

Reply:
Agreed. You either get List<>, which implements IList, or we synthesize the immutable collection, which should do the same.

Reply:
I think I agree, the synthesized type generated by the compiler should probably implement `IList` in addition to the generic interfaces. But it wasn't clear to me from the LDM notes whether the spec should guarantee that.

Reply:
that was my understanding of hte discussion and voting.  If we don't guarantee this, then what's the point?  Removing it would just break anyone who now takes a dependency on this.  And we want people to take a dependency on this so that they can use collection literals in these scenarios.

Reply:
Would it help to do a clarifying LDM email? I think we shouldn't conditionally light up IList depending on which of the interfaces is targeted. It's an interface dynamically discovered from `object`-typed instances, or sometimes `IEnumerable`-typed instances. The generic target interface type doesn't really have any bearing on the intended data binding scenario.

Reply:
Is supporting IList an existing requirement for regular types that users want to pass into data binding scenarios that are accepting IEnumerable<T>? Seems odd since you might be passing in a read-only collection that doesn't support adding. And you'll presumably end up with a less performance synthetic version

Maybe I've misunderstood and it's just saying to do that for synthetic versions of IList<T> and not IEnumerable<T>

Reply:
`IList` (and `IList<T>`) have `IsReadOnly` which is set to true in these scenarios. Part of consuming `IList`/`IList<T>` is knowing that that they will throw NotSupportedException from mutation methods if `IsReadOnly` is true.

In the case of Windows Forms, data binding doesn't recognize IEnumerable as a collection unless it finds `System.Collections.ICollection` and for other things, `System.Collections.IList`.